### DISCUSSION TITLE
Android 15 Framework Support

### DISCUSSION THREAD

**Discussion by codylund** [iBotPeaches/Apktool#3665] [Q&A] [answered]
With Android 15 GA on the horizon, can you provide an ETA for the corresponding framework support in ApkTool?

**Comment by iBotPeaches** [comment]
I'm basically waiting on a smali fix (the only hard-blocker I consider). The Android 15 aspect (with beta frameworks) appears to be fine.

I'm converting this to a discussion

**Reply by codylund** [reply]
Thanks, Connor! Do you have a rough ETA on that fix?

For transparency, my team pulls in Apktool as a critical dependency, and we are gearing up for Android 15. Any additional info is greatly appreciated.

**Reply by msft-cofitz** [reply]
echoing the above - will there be a new version of APKTool for supporting apps built targeting Android 15 / API 35? If yes, is there an ETA for that?

**Reply by iBotPeaches** [reply]
Yes I will release v2.10.x shortly after Android 15. This is under assumption that this smali regression is resolved. If not, I'll have to revert back to an older smali and suffer the faults it has - https://github.com/google/smali/pull/68

**Reply by msft-cofitz** [reply]
Is [Android 15 is released to AOSP](https://android-developers.googleblog.com/2024/09/android-15-is-released-to-aosp.html) the milestone are looking for from the Android side or is this pending a different item?

**Reply by iBotPeaches** [reply]
Yes I pulled the framework when that was released and updated it here - https://github.com/iBotPeaches/Apktool/commit/c5dbcd6532312b30a362d864ffa2390e7304af58

Smali still has a major regression. The fix is up, but unmerged and no release cut which is my blocker - https://github.com/google/smali/pull/72

**Reply by iBotPeaches** [reply]
Okay smali 3.0.8 is out. Will shore up some things and aim for a release soon.

**Reply by Tofalvy** [reply]
Any update on this?
I cannot see the 2.10.x release yet.

**Reply by iBotPeaches** [reply]
@Tofalvy why can't you see the 2.10.0 release?

**Reply by Tofalvy** [reply]
Based on the release description 2.10.0 doesn't support android 35 yet, it supports android API 34:
https://apktool.org/blog/apktool-2.10.0

That's why I thought that 2.10.x means x>0.

**Reply by iBotPeaches** [reply]
Hmm I guess I honestly switched up API34/35 in my head. Which means there is no release out that supports API 35 and meanwhile I'm working on 2.11.0 that supports API 36 - https://github.com/iBotPeaches/Apktool/commit/f85f9b3b60276394c29247047d7d75350e5e138b. So 35 was just missed.

So if you can run a bleeding edge build and just confirm API 35 works - I'll have to rush the timeline on API 2.11.0 or hotfix 2.10.x with whatever I need for API 35.

Build here - https://github.com/iBotPeaches/Apktool/actions/runs/12010277904/artifacts/2232878590

**Reply by starbucks2010** [reply]
Hi,

I decompiled/recompiled framework.jar from the S928B (S24 ultra) rooted with latest beta 3 OneUI7 Android 15 and flashed via twrp but the device doesn't boot. Let me know if any file/trace is needed.

I used apktool-v2.10.0-18-f8e1a596-SNAPSHOT.jar and the latest smali 3.0.9 20241219.

Thank you.

**Reply by iBotPeaches** [reply]
Yes please provide a trace, because 3.0.9 smali should have had the fix that I was waiting for. I was about to cut Apktool 2.11 this weekend, so to hear it doesn't work is frightening.

**Comment by starbucks2010** [comment]
[dropboxrc2.zip](https://github.com/user-attachments/files/18383021/dropboxrc2.zip)
here are the dropbox files from device
do you need also the framework-res.apk and framework.jar I used ?

**Reply by Emulond** [reply]
@starbucks2010 (Andrei), hello there! I'm a fellow custom user, such as you are.
Me and my dear friend ExtremeXT have stumbled upon your dropbox archive. We have noticed that there are native crashes in your logs array.
You see, we are investigating some tremendously complicated issue with OneUI runtime in [[OneUI] Native crashes #127](https://github.com/JingMatrix/LSPosed/issues/127), and we need some information about your current setup.
1. Which root solution do you use?
2. Do you use Zygisk? If so, which implementation?
3. Do you use LSPosed? If so, which implementation?
Best regards

**Reply by Emulond** [reply]
@starbucks2010, does this look amusing? Or do you know anything about the ethics? I have asked you a question you have explicitly ignored while making a new message. It wouldn't take you more than a minute. Do not be this dishonourable.

**Comment by starbucks2010** [comment]
I used batchapktool 3.8.0 with this config if it has any importance
![2025-01-11 00_40_37-BAT by BurSoft (Java 21 0 2 0 x64)](https://github.com/user-attachments/assets/ca7418af-5a77-488f-9a5b-3e9cd69a83e9)

**Reply by duhansysl** [reply]
is batchapktool working fine with new DEX stuff?

**Comment by starbucks2010** [comment]
@iBotPeaches any progress on the apktool supporting android 15?

**Reply by iBotPeaches** [reply]
I believe its going out tonight. I ran out of time last night.

**Reply by starbucks2010** [reply]
it will be a test version for testing android 15 compatibility ?

**Reply by iBotPeaches** [reply]
Its completed my basic tests so it will be an official release of 2.11.0

**Reply by ExtremeXT** [reply]
APKTool compiled from latest master branch successfully booted S24's OneUI 7 Beta 3 here.

**Comment by ExtremeXT** [comment]
Samsung OneUI 7 (based on Android 15, AP3A) still uses DEX version 039 in the header which APKTool doesn't seem to know how to handle right now. I'm guessing it should function just fine on AOSP or other OEM skins.

UN1CA method works for me: https://github.com/salvogiangri/UN1CA/commit/d3ad136068e7c77dea104f23784a965131ffff95

Using APKTool and Smali/BakSmali built from latest master branch in both sources. With this way I managed to boot just fine without removing invoke-custom calls in smali.

*JARs:*

Decompile with API 35 via APKTool, but with --no-src to not decompile DEX files:
```
java -jar apktool.jar decode --api-level 35 --no-debug-info --no-src framework.jar --output framework.jar.out
```

Decompile each DEX file via Google Smali, with API 29:
```
java -jar baksmali-3.0.9-fat.jar disassemble --api 29 --accessor-comments false --debug-info false --use-locals --output framework.jar.out\classes framework.jar.out\classes.dex
java -jar baksmali-3.0.9-fat.jar disassemble --api 29 --accessor-comments false --debug-info false --use-locals --output framework.jar.out\classes2 framework.jar.out\classes2.dex
java -jar baksmali-3.0.9-fat.jar disassemble --api 29 --accessor-comments false --debug-info false --use-locals --output framework.jar.out\classes3 framework.jar.out\classes3.dex
java -jar baksmali-3.0.9-fat.jar disassemble --api 29 --accessor-comments false --debug-info false --use-locals --output framework.jar.out\classes4 framework.jar.out\classes4.dex
java -jar baksmali-3.0.9-fat.jar disassemble --api 29 --accessor-comments false --debug-info false --use-locals --output framework.jar.out\classes5 framework.jar.out\classes5.dex
java -jar baksmali-3.0.9-fat.jar disassemble --api 29 --accessor-comments false --debug-info false --use-locals --output framework.jar.out\classes6 framework.jar.out\classes6.dex
```

Remove old DEX files:
```
del framework.jar.out\classes.dex
del framework.jar.out\classes2.dex
del framework.jar.out\classes3.dex
del framework.jar.out\classes4.dex
del framework.jar.out\classes5.dex
del framework.jar.out\classes6.dex
```

After changing the DEX files - compile them:
```
java -jar smali-3.0.9-fat.jar assemble --api 29 --output framework.jar.out\classes.dex framework.jar.out\classes
java -jar smali-3.0.9-fat.jar assemble --api 29 --output framework.jar.out\classes2.dex framework.jar.out\classes2
java -jar smali-3.0.9-fat.jar assemble --api 29 --output framework.jar.out\classes3.dex framework.jar.out\classes3
java -jar smali-3.0.9-fat.jar assemble --api 29 --output framework.jar.out\classes4.dex framework.jar.out\classes4
java -jar smali-3.0.9-fat.jar assemble --api 29 --output framework.jar.out\classes5.dex framework.jar.out\classes5
java -jar smali-3.0.9-fat.jar assemble --api 29 --output framework.jar.out\classes6.dex framework.jar.out\classes6
```

Compile JAR with APKTool:
```
java -jar apktool.jar build --copy-original framework.jar.out
```

Zipalign JAR:
```
zipalign -p 4 framework.jar.out\dist\framework.jar framework_mod.jar
```

*APKs: *

It's the same, just with the framework-res file loaded:

```
java -jar apktool.jar if -p res framework-res.apk
```

And add `--frame-path res` argument to all APKTool commands. You might also need `-no-res`.

Make sure oat/arm64/arm files in system/framework, along with boot-*.odex are removed. Push to phone, wipe dalvik cache and reboot.

Commands use Windows' PATH separator, but there's no reason for it to not work on Linux as well, provided that the path is adapted.

@starbucks2010 @iBotPeaches

**Comment by starbucks2010** [comment]
it seems quite a tedious job following this procedure but thank you for the details
I hope Connor will find a better solution for A15

**Reply by ExtremeXT** [reply]
Well, this isn't an APKTool issue per se, just Samsung being Samsung. But yes, it would be nice for APKTool to gain an official workaround.

**Comment by iBotPeaches** [comment, marked as answer]
With v2.11.0 release last night - I want to close this and track any regressions/further issues in their own tickets.
